File: app/src/main/java/com/elementary/tasks/core/utils/MemoryUtil.java
package com.elementary.tasks.core.utils;

import android.os.Environment;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.io.File;
import java.util.Locale;

/**
 * Copyright 2016 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

public class MemoryUtil {

    public static final String DIR_SD = "backup";
    private static final String DIR_IMAGE_CACHE = "img";
    private static final String DIR_PREFS = "preferences";
    public static final String DIR_NOTES_SD = "notes";
    public static final String DIR_GROUP_SD = "groups";
    public static final String DIR_BIRTHDAY_SD = "birthdays";
    public static final String DIR_PLACES_SD = "places";
    public static final String DIR_TEMPLATES_SD = "templates";
    private static final String DIR_MAIL_SD = "mail_attachments";
    private static final String DIR_SD_DBX_TMP = "tmp_dropbox";
    private static final String DIR_NOTES_SD_DBX_TMP = "tmp_dropbox_notes";
    private static final String DIR_GROUP_SD_DBX_TMP = "tmp_dropbox_groups";
    private static final String DIR_BIRTHDAY_SD_DBX_TMP = "tmp_dropbox_birthdays";
    private static final String DIR_PLACES_SD_DBX_TMP = "tmp_dropbox_places";
    private static final String DIR_TEMPLATES_SD_DBX_TMP = "tmp_dropbox_templates";
    private static final String DIR_PREFERENCES_SD_DBX_TMP = "tmp_dropbox_preferences";
    private static final String DIR_SD_GDRIVE_TMP = "tmp_gdrive";
    private static final String DIR_NOTES_SD_GDRIVE_TMP = "tmp_gdrive_notes";
    private static final String DIR_GROUP_SD_GDRIVE_TMP = "tmp_gdrive_group";
    private static final String DIR_BIRTHDAY_SD_GDRIVE_TMP = "tmp_gdrive_birthdays";
    private static final String DIR_PLACES_SD_GDRIVE_TMP = "tmp_gdrive_places";
    private static final String DIR_TEMPLATES_SD_GDRIVE_TMP = "tmp_gdrive_templates";
    private static final String DIR_PREFERENCES_SD_GDRIVE_TMP = "tmp_gdrive_preferences";

    private MemoryUtil() {
    }

    public static boolean isSdPresent() {
        String state = Environment.getExternalStorageState();
        return Environment.MEDIA_MOUNTED.equals(state) ||
                Environment.MEDIA_MOUNTED_READ_ONLY.equals(state);
    }

    @Nullable
    public static File getRemindersDir() {
        return getDir(DIR_SD);
    }

    @Nullable
    public static File getGroupsDir() {
        return getDir(DIR_GROUP_SD);
    }

    @Nullable
    public static File getBirthdaysDir() {
        return getDir(DIR_BIRTHDAY_SD);
    }

    @Nullable
    public static File getNotesDir() {
        return getDir(DIR_NOTES_SD);
    }

    @Nullable
    public static File getPlacesDir() {
        return getDir(DIR_PLACES_SD);
    }

    @Nullable
    public static File getTemplatesDir() {
        return getDir(DIR_TEMPLATES_SD);
    }

    @Nullable
    public static File getGooglePlacesDir() {
        return getDir(DIR_PLACES_SD_GDRIVE_TMP);
    }

    @Nullable
    public static File getGoogleTemplatesDir() {
        return getDir(DIR_TEMPLATES_SD_GDRIVE_TMP);
    }

    @Nullable
    public static File getDropboxPlacesDir() {
        return getDir(DIR_PLACES_SD_DBX_TMP);
    }

    @Nullable
    public static File getDropboxTemplatesDir() {
        return getDir(DIR_TEMPLATES_SD_DBX_TMP);
    }

    @Nullable
    public static File getDropboxRemindersDir() {
        return getDir(DIR_SD_DBX_TMP);
    }

    @Nullable
    public static File getDropboxGroupsDir() {
        return getDir(DIR_GROUP_SD_DBX_TMP);
    }

    @Nullable
    public static File getDropboxBirthdaysDir() {
        return getDir(DIR_BIRTHDAY_SD_DBX_TMP);
    }

    @Nullable
    public static File getDropboxNotesDir() {
        return getDir(DIR_NOTES_SD_DBX_TMP);
    }

    @Nullable
    public static File getGoogleRemindersDir() {
        return getDir(DIR_SD_GDRIVE_TMP);
    }

    @Nullable
    public static File getGoogleGroupsDir() {
        return getDir(DIR_GROUP_SD_GDRIVE_TMP);
    }

    @Nullable
    public static File getGoogleBirthdaysDir() {
        return getDir(DIR_BIRTHDAY_SD_GDRIVE_TMP);
    }

    @Nullable
    public static File getGoogleNotesDir() {
        return getDir(DIR_NOTES_SD_GDRIVE_TMP);
    }

    @Nullable
    public static File getMailDir() {
        return getDir(DIR_MAIL_SD);
    }

    @Nullable
    public static File getPrefsDir() {
        return getDir(DIR_PREFS);
    }

    @Nullable
    public static File getGooglePrefsDir() {
        return getDir(DIR_PREFERENCES_SD_GDRIVE_TMP);
    }

    @Nullable
    public static File getDropboxPrefsDir() {
        return getDir(DIR_PREFERENCES_SD_DBX_TMP);
    }

    @Nullable
    public static File getImageCacheDir() {
        return getDir(DIR_IMAGE_CACHE);
    }

    @Nullable
    public static File getParent() {
        return getDir("");
    }

    @Nullable
    private static File getDir(@NonNull String directory) {
        if (isSdPresent()) {
            File sdPath = Environment.getExternalStorageDirectory();
            File dir = new File(sdPath.toString() + "/JustReminder/" + directory);
            if (!dir.exists() && dir.mkdirs()) {
                return dir;
            }
            return dir;
        } else {
            return null;
        }
    }

    @Nullable
    public static File getImagesDir() {
        return getDir("image_cache");
    }

    @NonNull
    public static String humanReadableByte(long bytes, boolean si) {
        int unit = si ? 1000 : 1024;
        if (bytes < unit) {
            return bytes + " B";
        }
        int exp = (int) (Math.log(bytes) / Math.log(unit));
        String pre = (si ? "kMGTPE" : "KMGTPE").charAt(exp - 1) + (si ? "" : "");
        return String.format(Locale.getDefault(), "%.1f %sB", bytes / Math.pow(unit, exp), pre);
    }
}

File: app/src/main/java/com/elementary/tasks/core/utils/MemoryUtil.kt
package com.elementary.tasks.core.utils

import android.content.Context
import android.database.Cursor
import android.net.Uri
import android.os.Environment
import android.provider.OpenableColumns
import android.util.Base64
import android.util.Base64InputStream
import android.util.Base64OutputStream
import com.elementary.tasks.core.cloud.FileConfig
import com.elementary.tasks.core.data.models.Birthday
import com.elementary.tasks.core.data.models.NoteWithImages
import com.elementary.tasks.core.data.models.OldNote
import com.elementary.tasks.core.data.models.Place
import com.elementary.tasks.core.data.models.Reminder
import com.elementary.tasks.core.data.models.ReminderGroup
import com.elementary.tasks.core.data.models.SmsTemplate
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.google.gson.stream.JsonReader
import com.google.gson.stream.JsonWriter
import timber.log.Timber
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.FileWriter
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader
import java.io.OutputStream
import java.io.OutputStreamWriter
import java.nio.charset.StandardCharsets
import java.util.*
import kotlin.math.ln
import kotlin.math.pow


object MemoryUtil {

  private const val DIR_SD = "backup"
  private const val DIR_PREFS = "preferences"
  private const val DIR_NOTES_SD = "notes"
  private const val DIR_GROUP_SD = "groups"
  private const val DIR_BIRTHDAY_SD = "birthdays"
  private const val DIR_PLACES_SD = "places"
  private const val DIR_TEMPLATES_SD = "templates"

  private val isSdPresent: Boolean
    get() {
      val state = Environment.getExternalStorageState()
      return Environment.MEDIA_MOUNTED == state || Environment.MEDIA_MOUNTED_READ_ONLY == state
    }

  val remindersDir = getDir(DIR_SD)
  val groupsDir = getDir(DIR_GROUP_SD)
  val birthdaysDir = getDir(DIR_BIRTHDAY_SD)
  val notesDir = getDir(DIR_NOTES_SD)
  val placesDir = getDir(DIR_PLACES_SD)
  val templatesDir = getDir(DIR_TEMPLATES_SD)
  val prefsDir = getDir(DIR_PREFS)
  val parent = getDir("")
  val imagesDir = getDir("image_cache")

  private fun getDir(directory: String): File? {
    return if (isSdPresent) {
      val sdPath = Environment.getExternalStorageDirectory()
      val dir = File("$sdPath/JustReminder/$directory")
      if (!dir.exists() && dir.mkdirs()) {
        dir
      } else dir
    } else {
      null
    }
  }

  fun humanReadableByte(bytes: Long, si: Boolean): String {
    val unit = if (si) 1000 else 1024
    if (bytes < unit) {
      return "$bytes B"
    }
    val exp = (ln(bytes.toDouble()) / ln(unit.toDouble())).toInt()
    val pre = (if (si) "kMGTPE" else "KMGTPE")[exp - 1] + if (si) "" else ""
    return String.format(Locale.US, "%.1f %sB", bytes / unit.toDouble().pow(exp.toDouble()), pre)
  }

  @Throws(IOException::class)
  fun writeFileNoEncryption(file: File, data: String?): String? {
    if (data == null) return null
    try {
      val inputStream = ByteArrayInputStream(data.toByteArray())
      val buffer = ByteArray(8192)
      var bytesRead: Int
      val output = ByteArrayOutputStream()
      try {
        do {
          bytesRead = inputStream.read(buffer)
          if (bytesRead != -1) {
            output.write(buffer, 0, bytesRead)
          }
        } while (bytesRead != -1)
      } catch (e: IOException) {
        e.printStackTrace()
      }

      if (file.exists()) {
        file.delete()
      }
      val fw = FileWriter(file)
      fw.write(output.toString())
      fw.close()
      output.close()
    } catch (e: SecurityException) {
      return null
    }
    return file.toString()
  }

  fun <T> fromStream(stream: InputStream, clazz: Class<T>): T? {
    try {
      val output64 = Base64InputStream(stream, Base64.DEFAULT)
      val bufferedReader = BufferedReader(InputStreamReader(output64))
      val reader = JsonReader(bufferedReader)
      Timber.d("fromStream: $stream, $clazz")
      val t: T?
      try {
        t = Gson().fromJson<T>(reader, clazz)
      } catch (e: Exception) {
        return null
      } catch (e: OutOfMemoryError) {
        return null
      }
      reader.close()
      return t
    } catch (e: Exception) {
      e.printStackTrace()
      return null
    }
  }

  fun <T> fromStreamNoDecrypt(stream: InputStream, clazz: Class<T>): T? {
    try {
      val bufferedReader = BufferedReader(InputStreamReader(stream))
      val reader = JsonReader(bufferedReader)
      Timber.d("fromStream: $stream, $clazz")
      val t: T?
      try {
        t = Gson().fromJson<T>(reader, clazz)
      } catch (e: Exception) {
        return null
      } catch (e: OutOfMemoryError) {
        return null
      }
      reader.close()
      return t
    } catch (e: Exception) {
      e.printStackTrace()
      return null
    }
  }

  fun toStream(any: Any, outputStream: OutputStream): Boolean {
    try {
      System.gc()
      val output64 = Base64OutputStream(outputStream, Base64.DEFAULT)
      val bufferedWriter = BufferedWriter(OutputStreamWriter(output64, StandardCharsets.UTF_8))
      val writer = JsonWriter(bufferedWriter)
      val type = when (any) {
        is Reminder -> object : TypeToken<Reminder>() {}.type
        is Place -> object : TypeToken<Place>() {}.type
        is Birthday -> object : TypeToken<Birthday>() {}.type
        is ReminderGroup -> object : TypeToken<ReminderGroup>() {}.type
        is SmsTemplate -> object : TypeToken<SmsTemplate>() {}.type
        is NoteWithImages -> object : TypeToken<OldNote>() {}.type
        else -> null
      } ?: return false
      Timber.d("toStream: $type, $any")
      try {
        Gson().toJson(if (any is NoteWithImages) OldNote(any) else any, type, writer)
      } catch (e: Exception) {
        return false
      } catch (e: OutOfMemoryError) {
        System.gc()
        return false
      }
      writer.close()
      return true
    } catch (e: Exception) {
      e.printStackTrace()
      return false
    }
  }

  fun readFromUri(context: Context, uri: Uri, source: String = ""): Any? {
    val cr = context.contentResolver ?: return null
    var inputStream: InputStream? = null
    try {
      inputStream = cr.openInputStream(uri)
    } catch (e: Exception) {
      e.printStackTrace()
    }

    if (inputStream == null) {
      return null
    }
    val cursor: Cursor? = cr.query(uri, null, null,
      null, null, null)

    val name = try {
      cursor?.use {
        if (it.moveToFirst()) {
          val displayName: String =
            it.getString(it.getColumnIndex(OpenableColumns.DISPLAY_NAME))
          displayName
        } else {
          source
        }
      } ?: source
    } catch (e: Exception) {
      source
    }
    Timber.d("readFromUri: $name, $source")
    return try {
      val output64 = Base64InputStream(inputStream, Base64.DEFAULT)
      val r = JsonReader(BufferedReader(InputStreamReader(output64)))
      val type = when {
        name.endsWith(FileConfig.FILE_NAME_PLACE) -> {
          object : TypeToken<Place>() {}.type
        }
        name.endsWith(FileConfig.FILE_NAME_REMINDER) -> {
          object : TypeToken<Reminder>() {}.type
        }
        name.endsWith(FileConfig.FILE_NAME_BIRTHDAY) -> {
          object : TypeToken<Birthday>() {}.type
        }
        name.endsWith(FileConfig.FILE_NAME_GROUP) -> {
          object : TypeToken<ReminderGroup>() {}.type
        }
        name.endsWith(FileConfig.FILE_NAME_NOTE) -> {
          object : TypeToken<OldNote>() {}.type
        }
        name.endsWith(FileConfig.FILE_NAME_TEMPLATE) -> {
          object : TypeToken<SmsTemplate>() {}.type
        }
        else -> null
      }
      Gson().fromJson(r, type)
    } catch (e: Exception) {
      Timber.d("readFromUri: Bad JSON")
      e.printStackTrace()
      null
    }
  }
}
